include/swift/Sema/ConstraintLocator.h

```diff
@@ -85,6 +85,15 @@ enum ContextualTypePurpose : uint8_t {
   CTP_ComposedPropertyWrapper, ///< Composed wrapper type expected to match
                                ///< former 'wrappedValue' type
 
+  CTP_SingleValueStmtBranch, ///< The contextual type for a branch in a single
+                             ///< value statement expression.
+
+  CTP_SingleValueStmtBranchInSingleExprClosure, ///< The contextual type for a
+                                                ///< branch in a single value
+                                                ///< statement expression that's
+                                                ///< part of a single expression
+                                                ///< return in a closure.
```

Comment:
@xedin I don't love this, but it's needed to correctly handle the Void conversion for a single expression body closure. I don't suppose you can think of a better way of doing this? Ideally we'd just store a locator and get the information from there, but that might be a bit overkill.

Reply:
Have you considered this kind of a scheme:

a. allocate a single type variable that represents if/switch expression and just like with closures (and soon with taps) - provide a default type and delay until it either gets bound to contextual type or that default (default would just mean - proceed to step b.).

b. Once it's time to resolve, you can employ a trick similar to result builder transform and result type joining I implemented in (https://github.com/apple/swift/pull/62806):
  1. Tweak `TypeJoinExpr` to accept `SingleValueStmtExpr *` and generate a join based on branches (it can use `cs.getType` to extract type of each branch instead of walking into it).
  2. Use syntactic element generator to produce a conjunction and add `TypeJoinExpr` as the last element.
  3. Record conjunction as a new constraint.

This would mean that you don't have to allocate any extra variables, mess with brace elements and store contextual type information.